### PR TITLE
FIX - Fix batch provider, add multicall, fix context provider, dedup calls

### PR DESCRIPTION
**Pull request by halaprix**
# [Title](https://shortcutLink)

<please insert a shortcut link above>
  
## Changes 👷‍♀️
  <Please add short list of changes>
- item 1
  
## How to test 🧪
  <Please explain how to test your changes>
- step 1 ...

### CODE REVIEW

File: blockchain/config.ts [hunk 1]
```diff
@@ -61,6 +61,7 @@ const mainnetCacheUrl =
   process.env.MAINNET_CACHE_URL ||
   getConfig()?.publicRuntimeConfig?.mainnetCacheURL ||
   'https://oazo-bcache.new.oasis.app/api/v1'
+const mainnetRpc = `${process.env.ORIGIN ||  window.location.origin}/api/rpc?network=mainnet`
```
Comment (adamskrodzki) [inline blockchain/config.ts]: what about goerli ?

File: components/AppContext.ts [hunk 2]
```diff
@@ -489,11 +489,19 @@ export function setupAppContext() {
...
+    ),
+    shareReplay(1),
+  )
+
   const connectedContext$ = createContextConnected$(context$)
 
   combineLatest(account$, connectedContext$)
     .pipe(
       mergeMap(([account, network]) => {
+        console.log('account network', network.name, account?.toLowerCase())
```
Comment (adamskrodzki) [inline components/AppContext.ts:504]: remove console log

File: pages/api/rpc.tsx [hunk 3]
```diff
@@ -0,0 +1,107 @@
+import { withSentry } from '@sentry/nextjs'
+import axios from 'axios'
+import * as ethers from 'ethers'
+import { NextApiRequest, NextApiResponse } from 'next'
+
+function getRpcNode(network: string) {
+  switch (network) {
+    case 'mainnet':
```
Comment (adamskrodzki) [inline pages/api/rpc.tsx:8]: add goerli

**Review by adamskrodzki — CHANGES_REQUESTED** [review]

File: blockchain/jsonRpcBatchProvider.ts [hunk 4]
```diff
@@ -57,7 +63,7 @@ export class JsonRpcBatchProvider extends providers.JsonRpcProvider {
           request: deepCopy(request),
           provider: this,
         })
-
+        console.log('aggregator', batch ? batch.length : 0)
```
Comment (adamskrodzki) [inline blockchain/jsonRpcBatchProvider.ts:66]: remove console.log

**Review by adamskrodzki — APPROVED** [review]

File: pages/api/rpc.tsx [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    const hashedCalls = calls.map((call) =>
+      ethers.utils.keccak256(ethers.utils.toUtf8Bytes(`${call[0]}${call[1]}`)),
+    )
+
+    const dedupedCalls = calls.filter(
+      (call, index) =>
+        hashedCalls.indexOf(
+          ethers.utils.keccak256(ethers.utils.toUtf8Bytes(`${call[0]}${call[1]}`)),
+        ) === index,
+    )
```
Comment (paszkowskiDamian) [inline pages/api/rpc.tsx:76]: I think we could run hashing just once.

File: components/AppContext.ts [hunk 6]
```diff
@@ -563,16 +570,16 @@ export function setupAppContext() {
   const cdpManagerOwner$ = observe(onEveryBlock$, context$, cdpManagerOwner, bigNumberTostring)
   const cdpRegistryOwns$ = observe(onEveryBlock$, context$, cdpRegistryOwns)
   const cdpRegistryCdps$ = observe(onEveryBlock$, context$, cdpRegistryCdps)
```
Comment (paszkowskiDamian) [inline components/AppContext.ts:572]: I think those also can use chainContext

**Review by paszkowskiDamian — DISMISSED** [review]

**Review by adamskrodzki — DISMISSED** [review]

File: pages/api/rpc.tsx [hunk 7]
```diff
@@ -0,0 +1,116 @@
...
+interface CallWithHash {
+  hash: string
+  call: any
+}
+
+export async function rpc(req: NextApiRequest, res: NextApiResponse) {
+  let finalResponse: any[] = []
+  let mappedCalls: any[] = []
+  if (
+    process.env.ENABLE_MULTICALL &&
```
Comment (paszkowskiDamian) [inline pages/api/rpc.tsx:54]: You can remove it, it takes hours to set up. It would be faster to do redeploy

**Review by jakubswierczek — APPROVED** [review]

**Comment by halaprix** [comment]
🎉

**Pull request merged**